vendor/github.com/ethereum/go-ethereum/whisper/whisperv5/whisper.go

```diff
@@ -540,16 +542,48 @@ func (w *Whisper) Stop() error {
 	return nil
 }
 
+func waitTrusted(peer *Peer) error {
+	timeout := time.After(time.Minute)
+	ticker := time.NewTicker(2 * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-timeout:
+			return p2p.DiscTooManyPeers
+		case <-ticker.C:
+			if peer.Trusted() {
+				return nil
+			}
+		}
+	}
+}
+
 // HandlePeer is called by the underlying P2P layer when the whisper sub-protocol
 // connection is negotiated.
 func (wh *Whisper) HandlePeer(peer *p2p.Peer, rw p2p.MsgReadWriter) error {
+	var peersOverflow bool
+	wh.peerMu.RLock()
+	if int32(len(wh.peers)) > wh.maxPeers {
+		peersOverflow = true
+	}
+	wh.peerMu.RUnlock()
+
 	// Create the new peer and start tracking it
 	whisperPeer := newPeer(wh, peer, rw)
 
 	wh.peerMu.Lock()
 	wh.peers[whisperPeer] = struct{}{}
 	wh.peerMu.Unlock()
 
+	// we need to leave a window for a peer to become trusted
+	// otherwise we won't be able to connect with mailserver when max peers
+	// barier already reached
+	if peersOverflow {
+		if err := waitTrusted(whisperPeer); err != nil {
```

Comment:
This will block. Is that ok?

Reply:
yes, every handlePeer is spawned in a separate goroutine

Reply:
I see. Hm, still it bothers me that previously this method could return immediately and now it can block for 1 minute. It's a public method so it may influence callers if they do not run it in a separate goroutine.

Reply:
but HandlePeer was always blocking, see runMessageLoop at the bottom

Reply:
basically, this method is expected not to return immediately. i dont really know what was the purpose of making it public as well